ipie/addons/ephqmc/propagation/holstein.py

```diff
@@ -0,0 +1,182 @@
+        walkers.phia = propagate_one_body(walkers.phia, self.expH1[0])
+        if walkers.ndown > 0 and not walkers.rhf:
+            walkers.phib = propagate_one_body(walkers.phib, self.expH1[1])
+        synchronize()
+        self.timer.tgemm += time.time() - start_time
+
+    def propagate_phonons(self, walkers):
+        start_time = time.time()
+
+        pot = 0.25 * self.m * self.w0**2 * numpy.sum(walkers.x**2, axis=1)
+        pot = numpy.real(pot)
+        walkers.weight *= numpy.exp(-self.dt_ph * pot)
+
+        N = numpy.random.normal(loc=0.0, scale=self.scale, 
+                             size=(walkers.nwalkers, self.nsites))        
+        walkers.x = walkers.x + N 
+
+        pot = 0.25 * self.m * self.w0**2 * numpy.sum(walkers.x**2, axis=1)
+        pot = numpy.real(pot)
+        walkers.weight *= numpy.exp(-self.dt_ph * pot)
+            
+        walkers.weight *= numpy.exp(self.dt_ph * self.nsites * self.w0 / 2) #doesnt matter for estimators
+
+        synchronize()
+        self.timer.tgemm += time.time() - start_time
+
+    def propagate_electron(self, walkers, trial):
+        start_time = time.time()
+        ovlp = trial.calc_greens_function(walkers) 
+        synchronize()
+        self.timer.tgf += time.time() - start_time
+
+        expEph = numpy.exp(self.const * walkers.x)
+        
+        walkers.phia = propagate_one_body(walkers.phia, self.expH1[0])
+        walkers.phia = numpy.einsum('ni,nie->nie', expEph, walkers.phia)
+        walkers.phia = propagate_one_body(walkers.phia, self.expH1[0])
+        
+        if walkers.ndown > 0:
+            walkers.phib = propagate_one_body(walkers.phib, self.expH1[1])
+            walkers.phib = numpy.einsum('ni,nie->nie', expEph, walkers.phib)
+            walkers.phib = propagate_one_body(walkers.phib, self.expH1[1])
+
+    def propagate_walkers(self, walkers, hamiltonian, trial, eshift=None):
+        synchronize()
+        start_time = time.time()
+        ovlp = trial.calc_overlap(walkers)
+        synchronize()
+        self.timer.tgf += time.time() - start_time
+
+        # 2. Update Walkers
+        # 2.a DMC for phonon degrees of freedom
+        self.propagate_phonons(walkers)
+
+        # 2.b One-body propagation for electrons
+        self.propagate_electron(walkers, trial)
+
+        # 2.c DMC for phonon degrees of freedom
+        self.propagate_phonons(walkers)
+
+        # Update weights (and later do phaseless for multi-electron)
+        start_time = time.time()
+        ovlp_new = trial.calc_overlap(walkers)
+        synchronize()
+        self.timer.tovlp += time.time() - start_time
+
+        start_time = time.time()
+        self.update_weight(walkers, ovlp, ovlp_new)
+        synchronize()
+        self.timer.tupdate += time.time() - start_time
+
+    
+    def update_weight(self, walkers, ovlp, ovlp_new):
+        walkers.weight *= ovlp_new / ovlp
+
+
+class HolsteinPropagatorImportance(HolsteinPropagatorFree):
+    """"""
```

Comment:
Docstring.

Reply:
I would nearly just call this HolsteinPropagatorImportanceSampling or just HolsteinPropagator. Free projection is a rarer use case than phaseless.